src/contracts/libraries/MagnitudeCheckpoints.sol

```diff
@@ -0,0 +1,266 @@
+                low = mid + 1;
+            } else {
+                high = mid;
+            }
+        }
+        return high;
+    }
+
+    /**
+     * @dev Access an element of the array without performing bounds check. The position is assumed to be within bounds.
+     */
+    function _unsafeAccess(
+        MagnitudeCheckpoint[] storage self,
+        uint256 pos
+    ) private pure returns (MagnitudeCheckpoint storage result) {
+        assembly {
+            mstore(0, self.slot)
+            result.slot := add(keccak256(0, 0x20), pos)
+        }
+    }
+
+    /**
+     *
+     *                         ADDITIONAL FUNCTIONS FROM EIGEN-LABS
+     *
+     */
+
+    /**
+     * @dev Returns the value in the last (most recent) checkpoint with timestamp lower or equal than the search timestamp, or zero if there is none.
+     * In addition, returns the position of the checkpoint in the array.
+     */
+    function upperLookupWithPos(History storage self, uint32 timestamp) internal view returns (uint224, uint256) {
+        uint256 len = self._magnitudes.length;
+        uint256 pos = _upperBinaryLookup(self._magnitudes, timestamp, 0, len);
+        return pos == 0 ? (0, pos) : (_unsafeAccess(self._magnitudes, pos - 1)._value, pos - 1);
+    }
+
+    /**
+     * @dev Returns the value in the last (most recent) checkpoint with timestamp lower or equal than the search timestamp, or zero if there is none.
+     * In addition, returns the position of the checkpoint in the array.
+     *
+     * NOTE: This is a variant of {upperLookup} that is optimised to find "recent" checkpoint (checkpoints with high timestamps).
+     */
+    function upperLookupRecentWithPos(History storage self, uint32 timestamp) internal view returns (uint224, uint256) {
+        uint256 len = self._magnitudes.length;
+
+        uint256 low = 0;
+        uint256 high = len;
+
+        if (len > 5) {
+            uint256 mid = len - MathUpgradeable.sqrt(len);
+            if (timestamp < _unsafeAccess(self._magnitudes, mid)._timestamp) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        uint256 pos = _upperBinaryLookup(self._magnitudes, timestamp, low, high);
+
+        return pos == 0 ? (0, pos) : (_unsafeAccess(self._magnitudes, pos - 1)._value, pos - 1);
+    }
+
+    /// @notice WARNING: this function is only used because of the invariant property
+    /// that from the current timestamp, all future checkpointed magnitude values are strictly > current value.
+    /// Use function with extreme care for other situations.
+    function decrementCurrentAndFutureCheckpoints(
+        History storage self,
+        uint32 timestamp,
+        uint224 decrementValue
+    ) internal {
+        (uint224 value, uint256 pos) = upperLookupRecentWithPos(self, timestamp);
+
+        // if there is no checkpoint, return
+        pos == 0 ? type(uint256).max : pos;
```

Comment:
why does this mean there is no checkpoint?

Reply:
`upperLookupRecent` returns value of 0 if there is no checkpoint, this should probably be changed to check value == 0 instead because there could be a checkpoint value at pos 0.